### ISSUE TITLE
Wrong HLIL generated, missing pointer assignments

### ISSUE THREAD

**Issue by joelreymont** [Vector35/binaryninja-api#4322]
**Version and Platform (required):**
- Binary Ninja Version: 3.5.4286-dev, 58311b01
- OS: macos
- OS Version: 13.3
- CPU Architecture: arm64

The BN disassembly 

```
00048054  int32_t add_icc_handlers_48054(int32_t arg1, char* arg2)

00048054  30b5       push    {r4, r5, lr} {var_4} {__saved_r5} {__saved_r4}
00048056  cbb0       sub     sp, #0x12c
00048058  0546       mov     r5, r0
0004805a  0c46       mov     r4, r1
0004805c  3022       movs    r2, #0x30
0004805e  0021       movs    r1, #0
00048060  3ea8       add     r0, sp, #0xf8 {var_40}
00048062  7ff08afe   bl      #memset
00048066  7422       movs    r2, #0x74
00048068  0021       movs    r1, #0
0004806a  02a8       add     r0, sp, #8 {var_130}
0004806c  7ff085fe   bl      #memset
00048070  2f4b       ldr     r3, [pc, #0xbc]  {data_48130}  {icc_c6f68}
00048072  7c22       movs    r2, #0x7c
00048074  6946       mov     r1, sp {var_138}
00048076  0093       str     r3, [sp] {var_138}  {icc_c6f68}
00048078  2e4b       ldr     r3, [pc, #0xb8]  {data_48134}  {nullsub}
0004807a  0deb0200   add     r0, sp, r2 {var_bc}
0004807e  0193       str     r3, [sp, #4] {var_134}  {nullsub}
00048080  2d4b       ldr     r3, [pc, #0xb4]  {icc_alloc_cbuf_44d80}  {data_48138}
⚠️00048082  0493       str     r3, [sp, #0x10]  {icc_alloc_cbuf_44d80}
00048084  2d4b       ldr     r3, [pc, #0xb4]  {data_4813c}  {icc_c6b2e}
00048086  0593       str     r3, [sp, #0x14] {var_124}  {icc_c6b2e}
00048088  2d4b       ldr     r3, [pc, #0xb4]  {data_48140}  {icc_c6b40}
0004808a  0693       str     r3, [sp, #0x18] {var_120}  {icc_c6b40}
0004808c  2d4b       ldr     r3, [pc, #0xb4]  {icc_cbuf_alloc_impl}  {data_48144}
0004808e  0793       str     r3, [sp, #0x1c] {var_11c}  {icc_cbuf_alloc_impl}
00048090  2d4b       ldr     r3, [pc, #0xb4]  {icc_buf_44e34}  {data_48148}
00048092  0893       str     r3, [sp, #0x20] {var_118}  {icc_buf_44e34}
00048094  2d4b       ldr     r3, [pc, #0xb4]  {data_4814c}  {icc_c6b44}
00048096  0993       str     r3, [sp, #0x24] {var_114}  {icc_c6b44}
00048098  2d4b       ldr     r3, [pc, #0xb4]  {data_48150}  {icc_lock_48ad4}
0004809a  0a93       str     r3, [sp, #0x28] {var_110}  {icc_lock_48ad4}
0004809c  2d4b       ldr     r3, [pc, #0xb4]  {data_48154}  {nullsub}
0004809e  0b93       str     r3, [sp, #0x2c] {var_10c}  {nullsub}
000480a0  2d4b       ldr     r3, [pc, #0xb4]  {data_48158}  {icc_48b00}
000480a2  0c93       str     r3, [sp, #0x30] {var_108}  {icc_48b00}
000480a4  2d4b       ldr     r3, [pc, #0xb4]  {data_4815c}  {icc_48b10}
000480a6  0d93       str     r3, [sp, #0x34] {var_104}  {icc_48b10}
000480a8  2d4b       ldr     r3, [pc, #0xb4]  {data_48160}  {icc_488d8}
000480aa  0e93       str     r3, [sp, #0x38] {var_100}  {icc_488d8}
000480ac  2d4b       ldr     r3, [pc, #0xb4]  {data_48164}  {icc_c6e32}
000480ae  0f93       str     r3, [sp, #0x3c] {var_fc}  {icc_c6e32}
000480b0  2d4b       ldr     r3, [pc, #0xb4]  {data_48168}  {icc_c6de0}
000480b2  1093       str     r3, [sp, #0x40] {var_f8}  {icc_c6de0}
000480b4  2d4b       ldr     r3, [pc, #0xb4]  {data_4816c}  {icc_c6e02}
000480b6  1193       str     r3, [sp, #0x44] {var_f4}  {icc_c6e02}
000480b8  2d4b       ldr     r3, [pc, #0xb4]  {data_48170}  {icc_c6e5e}
000480ba  1293       str     r3, [sp, #0x48] {var_f0}  {icc_c6e5e}
000480bc  2d4b       ldr     r3, [pc, #0xb4]  {data_48174}  {icc_c6e68}
000480be  1393       str     r3, [sp, #0x4c] {var_ec}  {icc_c6e68}
000480c0  2d4b       ldr     r3, [pc, #0xb4]  {data_48178}  {icc_c6e68}
000480c2  1693       str     r3, [sp, #0x58] {var_e0}  {icc_c6e68}
000480c4  7ff02efe   bl      #memcpy
000480c8  0022       movs    r2, #0
000480ca  2c49       ldr     r1, [pc, #0xb0]  {data_4817c}  {data_5ffd0000}
000480cc  0323       movs    r3, #3
000480ce  8df81851   strb    r5, [sp, #0x118] {var_20}
000480d2  8df81721   strb    r2, [sp, #0x117] {var_21}  {0x0}
000480d6  adf81431   strh    r3, [sp, #0x114] {var_24}  {0x3}
000480da  294b       ldr     r3, [pc, #0xa4]  {data_48180}
000480dc  cde93f11   strd    r1, r1, [sp, #0xfc] {var_3c} {var_38}  {data_5ffd0000}  {data_5ffd0000}
000480e0  0968       ldr     r1, [r1]  {data_5ffd0000}
000480e2  9942       cmp     r1, r3
000480e4  4ff00101   mov     r1, #1
000480e8  0cbf       ite     eq
000480ea  8df81a21   strb    r2, [sp, #0x11a] {var_1e_1}  {0x0}
000480ee  2270       strb    r2, [r4]  {0x0}
000480f0  244a       ldr     r2, [pc, #0x90]  {data_48184}  {sub_48274}
000480f2  0cbf       ite     eq
000480f4  2170       strb    r1, [r4]  {0x1}
000480f6  8df81a11   strb    r1, [sp, #0x11a] {var_1e_2}  {0x1}
000480fa  012d       cmp     r5, #1
000480fc  4892       str     r2, [sp, #0x120] {var_18}  {sub_48274}
000480fe  4ff00002   mov     r2, #0
00048102  4992       str     r2, [sp, #0x124] {var_14}  {0x0}
00048104  0bd1       bne     #0x4811e

00048106  2049       ldr     r1, [pc, #0x80]  {data_48188}  {data_5ffc0000}
00048108  cde93f11   strd    r1, r1, [sp, #0xfc] {var_3c_1} {var_38_1}  {data_5ffc0000}  {data_5ffc0000}
0004810c  0968       ldr     r1, [r1]  {data_5ffc0000}
0004810e  9942       cmp     r1, r3
00048110  07bf       ittee   eq
00048112  8df81a21   strb    r2, [sp, #0x11a] {var_1e_3}  {0x0}
00048116  2570       strb    r5, [r4]
00048118  8df81a51   strb    r5, [sp, #0x11a] {var_1e_4}
0004811c  2270       strb    r2, [r4]  {0x0}

0004811e  1fa8       add     r0, sp, #0x7c {var_bc}
00048120  fdf726fd   bl      #sub_45b70
00048124  0038       subs    r0, #0
00048126  18bf       it      ne
00048128  0120       movs    r0, #1
0004812a  4042       rsbs    r0, r0, #0
0004812c  4bb0       add     sp, #0x12c
0004812e  30bd       pop     {r4, r5, pc} {__saved_r4} {__saved_r5} {var_4}

00048130  void* data_48130 = icc_c6f68
00048134  void* data_48134 = nullsub
00048138  void* data_48138 = icc_alloc_cbuf_44d80
0004813c  void* data_4813c = icc_c6b2e
00048140  void* data_48140 = icc_c6b40
00048144  void* data_48144 = icc_cbuf_alloc_impl
00048148  void* data_48148 = icc_buf_44e34
0004814c  void* data_4814c = icc_c6b44
00048150  void* data_48150 = icc_lock_48ad4
00048154  void* data_48154 = nullsub
00048158  void* data_48158 = icc_48b00
0004815c  void* data_4815c = icc_48b10
00048160  void* data_48160 = icc_488d8
00048164  void* data_48164 = icc_c6e32
00048168  void* data_48168 = icc_c6de0
0004816c  void* data_4816c = icc_c6e02
00048170  void* data_48170 = icc_c6e5e
00048174  void* data_48174 = icc_c6e68
00048178  void* data_48178 = icc_c6e68
0004817c  int32_t data_4817c = 0x5ffd0000
00048180  void* data_48180 = 0xdeadbeef
00048184  void* data_48184 = sub_48274
00048188  int32_t data_48188 = 0x5ffc0000
```

Expected code (IDA)

```
int __fastcall add_icc_handlers_48054(int a1, _BYTE *a2)
{
  bool v4; // zf
  int v5; // r0
  int v7[31]; // [sp+0h] [bp-138h] BYREF
  char v8[124]; // [sp+7Ch] [bp-BCh] BYREF
  char v9[48]; // [sp+F8h] [bp-40h] BYREF

  memset(v9, 0, sizeof(v9));
  memset(&v7[2], 0, 0x74u);
  v7[0] = icc_C6EF8;
  v7[1] = nullsub_5;
  v7[4] = icc_alloc_cbuf_44D80;
  v7[5] = icc_C6B2E;
  v7[6] = icc_C6B40;
  v7[7] = icc_cbuf_alloc_impl;
  v7[8] = icc_buf_44E34;
  v7[9] = icc_C6B44;
  v7[10] = icc_lock_48AD4;
  v7[11] = nullsub_26;
  v7[12] = icc_48B00;
  v7[13] = icc_48B10;
  v7[14] = icc_488D8;
  v7[15] = icc_C6E32;
  v7[16] = icc_C6DE0;
  v7[17] = icc_C6E02;
  v7[18] = icc_C6E5E;
  v7[19] = icc_C6E68;
  v7[22] = icc_C6EA0;
  memcpy(v8, v7, sizeof(v8));
  v9[32] = a1;
  v9[31] = 0;
  *&v9[28] = 3;
  *&v9[4] = dword_5FFD0000;
  *&v9[8] = dword_5FFD0000;
  v4 = dword_5FFD0000[0] == 0xDEADBEEF;
  if ( dword_5FFD0000[0] == 0xDEADBEEF )
    v9[34] = 0;
  else
    *a2 = 0;
  if ( v4 )
    *a2 = 1;
  else
    v9[34] = 1;
  *&v9[40] = add_icc_event_task;
  *&v9[44] = 0;
  if ( a1 == 1 )
  {
    *&v9[4] = dword_5FFC0000;
    *&v9[8] = dword_5FFC0000;
    if ( dword_5FFC0000[0] == 0xDEADBEEF )
    {
      v9[34] = 0;
      *a2 = 1;
    }
    else
    {
      v9[34] = 1;
      *a2 = 0;
    }
  }
  sub_45B70(v8);
  return -(v5 != 0);
}
```

Wrong HLIL that's missing the function pointer assignments 

```
00048054  int32_t add_icc_handlers_48054(int32_t arg1, char* arg2)

00048062      void var_40
00048062      memset(&var_40, 0, 0x30)
0004806c      void var_130
0004806c      memset(&var_130, 0, 0x74)
00048076      void* const var_138 = icc_c6f68
0004807e      void* const var_134 = nullsub
⚠️00048082      int32_t var_128
00048082      __builtin_memcpy(var_128, "\x81M\x04\x00/k\x0c\x00Ak\x0c\x00\xbdM\x04\x005N\x04\x00Ek\x0c\x00\xd5\x8a\x04\x00\xf7n\x0c\x00\x01\x8b\x04\x00\x11\x8b\x04\x00\xd9\x88\x04\x003n\x0c\x00\xe1m\x0c\x00\x03n\x0c\x00_n\x0c\x00in\x0c\x00", 0x40)
000480c2      void* const var_e0 = icc_c6e68
000480c4      void var_bc
000480c4      memcpy(&var_bc, &var_138, &data_7c)
000480ce      char var_20 = arg1.b
000480d2      char var_21 = 0
000480d6      int16_t var_24 = 3
000480dc      int32_t var_3c = 0x5ffd0000
000480dc      int32_t var_38 = 0x5ffd0000
000480e0      int32_t r1_1 = data_5ffd0000
000480ea      if (r1_1 == 0xdeadbeef)
000480ea          char var_1e_1 = 0
000480ee      if (r1_1 != 0xdeadbeef)
000480ee          *arg2 = 0
000480f4      if (r1_1 == 0xdeadbeef)
000480f4          *arg2 = 1
000480f6      if (r1_1 != 0xdeadbeef)
000480f6          char var_1e_2 = 1
000480fc      void* const var_18 = sub_48274
00048102      int32_t var_14 = 0
00048104      if (arg1 == 1)
00048108          int32_t var_3c_1 = 0x5ffc0000
00048108          int32_t var_38_1 = 0x5ffc0000
0004810c          int32_t r1_2 = data_5ffc0000
00048112          if (r1_2 == 0xdeadbeef)
00048112              char var_1e_3 = 0
00048116          if (r1_2 == 0xdeadbeef)
00048116              *arg2 = arg1.b
00048118          if (r1_2 != 0xdeadbeef)
00048118              char var_1e_4 = arg1.b
0004811c          if (r1_2 != 0xdeadbeef)
0004811c              *arg2 = 0
00048120      int32_t r0_4 = sub_45b70(&var_bc)
00048124      int32_t r0_5 = r0_4
00048128      if (r0_4 != 0)
00048128          r0_5 = 1
0004812e      return 0 - r0_5
```

Anyone from V35 should search for "Encourage Salesman Prompt Delay" to find the database.

**Comment by joelreymont** [comment]
Same issue here 

```
0004d95c  int32_t egm7_backup_layout_init()

0004d95c  0e4a       ldr     r2, [pc, #0x38]  {data_4d998}
0004d95e  0821       movs    r1, #8
0004d960  0e4b       ldr     r3, [pc, #0x38]  {data_4d99c}
0004d962  00b5       push    {lr} {var_4}
0004d964  91b0       sub     sp, #0x44
0004d966  9b1a       subs    r3, r3, r2  {0x1800}
0004d968  6846       mov     r0, sp {var_48}
⚠️0004d96a  0092       str     r2, [sp] {var_48}  {0x2007e000}
0004d96c  0f93       str     r3, [sp, #0x3c] {var_c}  {0x1800}
0004d96e  cde90132   strd    r3, r2, [sp, #4] {var_44} {var_40}  {0x1800}  {0x2007e000}
0004d972  cde90332   strd    r3, r2, [sp, #0xc] {var_3c} {var_38}  {0x1800}  {0x2007e000}
0004d976  cde90532   strd    r3, r2, [sp, #0x14] {var_34} {var_30}  {0x1800}  {0x2007e000}
0004d97a  cde90732   strd    r3, r2, [sp, #0x1c] {var_2c} {var_28}  {0x1800}  {0x2007e000}
0004d97e  cde90932   strd    r3, r2, [sp, #0x24] {var_24} {var_20}  {0x1800}  {0x2007e000}
0004d982  cde90b32   strd    r3, r2, [sp, #0x2c] {var_1c} {var_18}  {0x1800}  {0x2007e000}
0004d986  cde90d32   strd    r3, r2, [sp, #0x34] {var_14} {var_10}  {0x1800}  {0x2007e000}
0004d98a  e8f759ff   bl      #sub_36840
0004d98e  0020       movs    r0, #0
0004d990  11b0       add     sp, #0x44
0004d992  5df804fb   pop     {pc} {var_4}
```

vs

```
0004d95c  int32_t egm7_backup_layout_init()

⚠️0004d96a      int32_t var_48
0004d96a      __builtin_memcpy(var_48, "\x00\xe0\x07 \x00\x18\x00\x00\x00\xe0\x07 \x00\x18\x00\x00\x00\xe0\x07 \x00\x18\x00\x00\x00\xe0\x07 \x00\x18\x00\x00\x00\xe0\x07 \x00\x18\x00\x00\x00\xe0\x07 \x00\x18\x00\x00\x00\xe0\x07 \x00\x18\x00\x00\x00\xe0\x07 \x00\x18\x00\x00", 0x40)
0004d98a      sub_36840(&var_48, 8)
0004d992      return 0
```

**Comment by plafosse** [comment]
The issue here is three fold. 
First we're not recognizing these constants as pointers. https://github.com/Vector35/binaryninja-api/issues/4399
Second issue is we're coercing a set of contiguous stack writes to a builtin_memcpy. We are strongly reconsidering as it seems to be getting in the way more than its helping. https://github.com/Vector35/binaryninja-api/issues/4398
Third issue, we're not recognizing stack arrays. This in particular is a tricky problem that we haven't gotten around to yet. https://github.com/Vector35/binaryninja-api/issues/2570
Closing this issue as a duplicate even though I had to create two issues to cover this one.
